File: atomic-services/etaCalculator/internal/haversine/haversine.go
package haversine

import "math"

// Converts degrees(coordinates) to radians
func toRad(deg float64) float64 {
	return deg * math.Pi / 180
}

// DistanceKm manually calculates the distance between the two points
// It uses the haversine formula, which takes in the lat, long values on a globe to calc the distance
func DistanceKm(lat1, lng1, lat2, lng2 float64) float64 {
	const R = 6371.0
	dLat := toRad(math.Abs(lat1 - lat2)) //since Singapore is above the equator, just have to minus
	dLng := toRad(math.Abs(lng1 - lng2)) //same hemisphere as well
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(toRad(lat1))*math.Cos(toRad(lat2))*
			math.Sin(dLng/2)*math.Sin(dLng/2)
	return R * 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

// FallbackETA	estimates the time that the driver will take based off distance
// Assumes 30km/hr speed on road, slower than irl because we are using straight line distance
func FallbackETA(lat1, lng1, lat2, lng2 float64) (minutes int, distMeters int) {
	distKm := DistanceKm(lat1, lng1, lat2, lng2)
	minutes = int(math.Ceil(distKm / 30.0 * 60.0))
	if minutes < 1 {
		minutes = 1
	}
	return minutes, int(distKm * 1000)
}

//Euclidean kinda works especially for Singapore, but its meant for flat planes only
//func EuclideanDistance(lat1, lng1, lat2, lng2 float64) float64 {
//	dLat := math.Pow(lat1-lat2, 2.0) * 111.0
//	dlng := math.Pow(lng1-lng2, 2.0) * 111.0
//	return math.Sqrt(dLat + dlng)
//}

File: atomic-services/kitchen-assignment/haversine.py
import math

def to_rad(deg: float) -> float:
    """Converts degrees to radians"""
    return deg * math.pi / 180

def distance_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """
    Haversine distance between two lat/lng points in km.
    """
    R = 6371.0  # Earth radius km
    d_lat = to_rad(abs(lat1 - lat2))
    d_lng = to_rad(abs(lng1 - lng2))
    a = (math.sin(d_lat / 2) ** 2 +
         math.cos(to_rad(lat1)) * math.cos(to_rad(lat2)) *
         math.sin(d_lng / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c

def estimate_duration_seconds(dist_km: float) -> int:
    """
    Conservative ETA at 30 km/h avg speed.
    """
    duration_h = dist_km / 30
    duration_sec = duration_h * 3600
    return max(int(math.ceil(duration_sec)), 60)  # Min 1 min

def distance_result(dist_km: float) -> dict:
    """Mock DistanceResult for compatibility."""
    dist_m = dist_km * 1000
    dur_s = estimate_duration_seconds(dist_km)
    return {'distance_meters': dist_m, 'duration_seconds': dur_s, 'status': 'OK'}

